core/services/workflows/workflow.go

```diff
@@ -9,15 +18,170 @@ type Capability struct {
 	Config map[string]any `yaml:"config"`
 }
 
-type Workflow struct {
+type workflowSpec struct {
 	Triggers  []Capability `yaml:"triggers"`
 	Actions   []Capability `yaml:"actions"`
 	Consensus []Capability `yaml:"consensus"`
 	Targets   []Capability `yaml:"targets"`
 }
 
-func Parse(yamlWorkflow string) (*Workflow, error) {
-	wf := &Workflow{}
-	err := yaml.Unmarshal([]byte(yamlWorkflow), wf)
+func (w *workflowSpec) steps() []Capability {
+	s := []Capability{}
+	s = append(s, w.Actions...)
+	s = append(s, w.Consensus...)
+	s = append(s, w.Targets...)
+	return s
+}
+
+type workflow struct {
+	graph.Graph[string, *node]
+
+	triggers []*triggerCapability
+
+	spec *workflowSpec
+}
+
+func (w *workflow) walkDo(start string, do func(n *node) error) error {
+	var outerErr error
+	err := graph.BFS(w.Graph, start, func(ref string) bool {
+		n, err := w.Graph.Vertex(ref)
+		if err != nil {
+			outerErr = err
+			return true
+		}
+
+		err = do(n)
+		if err != nil {
+			outerErr = err
+			return true
+		}
+
+		return false
+	})
+	if err != nil {
+		return err
+	}
+
+	return outerErr
+}
+
+func (w *workflow) adjacentNodes(start string) ([]*node, error) {
+	nodes := []*node{}
+	m, err := w.Graph.AdjacencyMap()
+	if err != nil {
+		return nil, err
+	}
+
+	adj, ok := m[start]
+	if !ok {
+		return nil, fmt.Errorf("could not find node with ref %s", start)
+	}
+
+	for adjacentRef := range adj {
+		n, err := w.Graph.Vertex(adjacentRef)
+		if err != nil {
+			return nil, err
+		}
+
+		nodes = append(nodes, n)
+	}
+
+	return nodes, nil
+}
+
+type node struct {
+	Capability
+	dependencies     []string
+	cachedCapability capabilities.CallbackExecutable
+	cachedConfig     *values.Map
+}
+
+type triggerCapability struct {
+	Capability
+	cachedTrigger capabilities.TriggerCapability
+}
+
+const (
+	keywordTrigger = "trigger"
+)
+
+func Parse(yamlWorkflow string) (*workflow, error) {
+	wfs := &workflowSpec{}
+	err := yaml.Unmarshal([]byte(yamlWorkflow), wfs)
+	if err != nil {
+		return nil, err
+	}
+
+	// Construct and validate the graph. We instantiate an
+	// empty graph with just one starting entry: `trigger`.
+	// This provides the starting point for our graph and
+	// points to all dependent nodes.
+	nodeHash := func(n *node) string {
+		return n.Ref
+	}
+	g := graph.New(
+		nodeHash,
+		graph.PreventCycles(),
+		graph.Directed(),
+	)
+	err = g.AddVertex(&node{
+		Capability: Capability{Ref: keywordTrigger},
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	// Next, let's populate the other entries in the graph.
+	// We allow the `Ref` field to be empty, but in that case
+	// default the `Ref` to the type.
```

Comment:
Hmmmm... the spec doesn't always require ref to be present, eg. for targets, which is why I added this.

I'll mark this as a TODO since I agree this probably isn't right, but we do need a way to universally identify a node within a workflow in cases where a ref isn't present though, as it's required by the graph library.